File: scrabble.cpp
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

int whoWins(vector<vector<int>> trie, vector<bool> isWord, int curNode, vector<bool> letters) { //0 is draw, 1 is win, and 2 is lose
    if(isWord[curNode]) //other player made this word, so you lose
        return 2;
    bool canDraw = false;
    for(int i = 0; i < 26; i++) { //iterate through all the letters
        if(!letters[i]) continue; //if you can't place this letter, continue
        if(trie[curNode][i] == 0) { //placing this letter doesn't lead to any other words, so it causes a draw
            canDraw = true;
            continue;
        }
        int res = whoWins(trie, isWord, trie[curNode][i], letters); //try placing this letter
        if(res == 0) canDraw = true; //if your opponent draws when placing this letter, then you can draw by placing this letter
        if(res == 2) return 1; //if your opponent loses after placing this letter, you win
    }
    //if you got past the for loop, that means you can't win
    if(canDraw) return 0; //if you can draw, do that
    return 2; //else you lose
}

int main() {
    int g;
    cin >> g;
    int c = 1;

    while(g-->0) {
        int n, m, count = 26;
        cin >> n >> m;
        char word[10001];
        cin >> word;
        vector<bool> letters(26); //says if a letter is used

        for(int i = 0; i < strlen(word); i++) letters[word[i]-'a'] = true;

        vector<vector<int>> trie(200001); //gives next node in trie
        vector<bool> isWord(200001); //checks to see if node in trie is a word
        for(int i = 0; i < trie.size(); i++) //initialize trie
            trie[i] = vector<int>(26);
        int nextAvail = 1; //next available node in trie
        
        for(int i = 0; i < m; i++) {
            cin >> word;
            int curNode = 0;
            int len = strlen(word);
            for(int j = 0; j < len; j++) { //insert this word into the trie
                if(trie[curNode][word[j]-'a'] == 0)
                    trie[curNode][word[j]-'a'] = nextAvail++;
                curNode = trie[curNode][word[j]-'a'];
            }
            isWord[curNode] = true;
        }
        int res = whoWins(trie, isWord, 0, letters);
        if(res == 0)
            cout << "Game #" << c++ << ": Draw" << endl;
        else if(res == 1)
            cout << "Game #" << c++ << ": Alice wins" << endl;
        else
            cout << "Game #" << c++ << ": Bob wins" << endl;
    }

    return 0;
}

File: scrabble.py
# solution to: Super Scrabble for HSPT Online 2019
# author: John Edwards 2019-04-19

import sys
sys.setrecursionlimit(100001)

is_word = '_end_'
def make_trie(lexicon):
    root = dict()
    for word in lexicon:
        node = root
        for letter in word:
            node = node.setdefault(letter, {})
        node[is_word] = is_word
    return root

# dfs down the trie
def dfs(p, node, letters):
    # if the word that has been placed down is valid
    # then player p loses, so return 1-p
    if is_word in node:
        return 1-p

    # set flags to False
    win = False
    draw = False
    # try placing every valid letter
    for c in letters:
        # if this letter is part of a word
        # then dfs from that node as the other player
        if c in node:
            # recurse down this path of the trie
            sub = dfs(1-p, node[c], letters)
            # does choosing the letter cause you to win?
            if sub == p:
                win = True
            # if you don't win, do you draw?
            elif sub == 2:
                draw = True
        # if this letter makes an invalid word
        # then placing it forces a draw
        else:
            draw = True
    
    # if any letter choice forced you to win you win
    if win:
        return p
    # otherwise if you can force a draw return 2
    elif draw:
        return 2
    # otherwise you lose
    else:
        return 1-p

t = int(input())
for tt in range(1, t+1):
    n, m = map(int, input().split())
    letters = input()
    lexicon = []
    for i in range(m):
        lexicon.append(input())
    
    # make trie based off of lexicon
    trie = make_trie(lexicon)
    
    out = ["Alice wins", "Bob wins", "Draw"]
    print("Game #%d: %s" % (tt, out[dfs(0, trie, letters)]))